src/storage/txn/scheduler.rs

```diff
@@ -913,40 +958,73 @@ impl Scheduler {
     /// the method initiates a get snapshot operation for furthur processing.
-    fn lock_and_get_snapshot(&mut self, cid: u64) {
+    fn lock_and_register_get_snapshot(&mut self, cid: u64) {
         if self.acquire_lock(cid) {
-            self.get_snapshot(cid);
+            let ctx = self.extract_context(cid).clone();
+            let mut group = self.grouped_cmds
+                .entry(HashableContext(ctx))
+                .or_insert_with(Vec::new);
+            group.push(cid);
         }
     }
 
     pub fn run(&mut self, receiver: Receiver<Msg>) -> Result<()> {
+        let mut msgs = vec![];
         loop {
             let msg = box_try!(receiver.recv());
-            match msg {
-                Msg::Quit => return Ok(()),
-                Msg::RawCmd { cmd, cb } => self.on_receive_new_cmd(cmd, cb),
-                Msg::SnapshotFinished { cid, cb_ctx, snapshot } => {
-                    self.on_snapshot_finished(cid, cb_ctx, snapshot)
+            msgs.push(msg);
+            loop {
+                match receiver.try_recv() {
+                    Ok(msg) => {
+                        msgs.push(msg);
+                        if msgs.len() >= CMD_BATCH_SIZE {
+                            break;
+                        }
+                    }
+                    Err(TryRecvError::Empty) => break,
+                    _ => (),
                 }
-                Msg::ReadFinished { cid, pr } => self.on_read_finished(cid, pr),
-                Msg::WritePrepareFinished { cid, cmd, pr, to_be_write } => {
-                    self.on_write_prepare_finished(cid, cmd, pr, to_be_write)
+            }
+
+            for msg in msgs.drain(..) {
+                match msg {
+                    Msg::Quit => return Ok(()),
+                    Msg::RawCmd { cmd, cb } => self.on_receive_new_cmd(cmd, cb),
+                    Msg::SnapshotFinished { cid, cb_ctx, snapshot } => {
+                        self.on_snapshot_finished(cid, cb_ctx, snapshot)
+                    }
+                    Msg::ReadFinished { cid, pr } => self.on_read_finished(cid, pr),
+                    Msg::WritePrepareFinished { cid, cmd, pr, to_be_write } => {
+                        self.on_write_prepare_finished(cid, cmd, pr, to_be_write)
+                    }
+                    Msg::WritePrepareFailed { cid, err } => self.on_write_prepare_failed(cid, err),
+                    Msg::WriteFinished { cid, pr, result, .. } => {
+                        self.on_write_finished(cid, pr, result)
+                    }
                 }
-                Msg::WritePrepareFailed { cid, err } => self.on_write_prepare_failed(cid, err),
-                Msg::WriteFinished { cid, pr, result, .. } => {
-                    self.on_write_finished(cid, pr, result)
+            }
+
+            if !self.grouped_cmds.is_empty() {
+                let m = self.grouped_cmds.clone();
+                for (ctx, cids) in m {
+                    if !cids.is_empty() {
```

Comment:
you should add a histogram metric to check how many items in one group.
If we have only few one, it is no benefit to do this.